Index/pinot-label.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.36.
.TH PINOT-LABEL "1" "November 2007" "pinot-label - pinot 0.80" "User Commands"
.SH NAME
pinot-label \- Label files from the command-line
.SH SYNOPSIS
.B pinot-label
[\fIOPTIONS\fR] [\fIFILES\fR]
.SH DESCRIPTION
pinot\-label \- Label files from the command\-line
.SH OPTIONS
.TP
\fB\-g\fR, \fB\-\-get\fR
get the labels list for the given file
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-l\fR, \fB\-\-list\fR
list known labels
.TP
\fB\-s\fR, \fB\-\-set\fR
set labels on the given file
.TP
\fB\-v\fR, \fB\-\-version\fR
output version information and exit
.SH EXAMPLES
pinot\-label \fB\-\-get\fR /home/fabrice/Documents/Bozo.txt
.PP
pinot\-label \fB\-\-list\fR
.PP
pinot\-label \fB\-\-set\fR "[Clowns][Fun][My Hero]" /home/fabrice/Documents/Bozo.txt
.SH "REPORTING BUGS"
Report bugs to fabricecolin@users.berlios.de
.PP
This is free software.  You may redistribute copies of it under the terms of
the GNU General Public License <http://www.gnu.org/licenses/old\-licenses/gpl\-2.0.html>.
There is NO WARRANTY, to the extent permitted by law.


NEWS
2007/11/01 version_0_8_0
Build :
 - SMP builds, thanks to Gabriel C
Index :
 - unknown document types can be indexed if one of their parent types is known
 - new pinot-label tool to get, set and list labels on indexed files from the
  command-line
 - limit external programs to 5 minutes of CPU time
Search :
 - support for date (year, month, day), time (hours, minutes, seconds) and
  size (in bytes) ranges
 - attempt to correct the spelling of index queries that don't match anything
 - log how long queries take
 - keep connection to remote databases alive
 - better query expansion
 - the Yahoo! plugin was replaced with the Yahoo! API plugin
 - pass queries to Web engines unmodified, without attempting to filter results
  based on a filter or a range used in the query
Daemon :
 - fixed Reload method
 - new D-Bus methods to manage labels
 - new --reindex option
UI :
 - Search This For menu to search in results
 - suggest spelling corrections for index queries that don't match anything
 - all indices can be browsed
 - inline URL completion in the import dialog box, based on previous results
 - better support for user-specific MIME settings, thanks to Lee Marks
 - reload MIME settings when they are edited
 - results of stored queries can be sorted by relevance (default) or by date
 - new History button to show previous results for a stored query
 - documents' properties are updated in the background
 - documents' terms can be saved to a file
 - Status window shows description of errors
 - send a Reload to the daemon only when the relevant preferences are modified
 - updated Dutch translation by JW
 - updated Brazilian Portuguese translation by Leonardo Melo

2007/08/23 version_0_7_6
Build :
 - also look for textcat.h in libtextcat
Monitor :
 - a deletion would deadlock the monitor and prevent from processing any
  further event
 - unindex directories' contents when deleted and update when moved
Collect :
 - try to open files with NO_ATIME if possible
 - Neon-based downloader had not been brought up to date
Index :
 - replaced na(t)ive tokenizer with Xapian's TermGenerator. No effort is
  made to convert text to UTF-8 yet, so this depends on document formats
  and encodings
 - preliminary support for spelling corrections, without user feedback
 - index directories are tagged as cache directories that they are skipped
  by "tar --exclude-caches"
 - preserve documents' title if possible, use user-specified title on import
Search :
 - fixed Sherlock parser for boost 1.34
 - support for gSOAP 2.7.9e
 - when searching an index, don't resort to OR'ing all terms if the original
  query doesn't match anything, this only confused users
Deskbar :
 - removed unnecessary shebang in script
Daemon :
 - new D-Bus method Reload, that enables to reload the configuration and act
  upon it whenever modified by the UI
 - SimpleQuery still resorts to OR'ing all terms if the original query doesn't
  match anything
UI :
 - fixed some minor cosmetic bugs, tweaked a few things
 - user-specific MIME settings in ~/.local have priority over system settings
 - queries with at least a start date can be run, and filter a documents list
 - the label specified on import was ignored
 - new traditional Chinese translation by Yung-Chung Lin
 - updated Portuguese translation by _PN_boy
 - updated Swedish translation by Daniel Nylander

2007/07/28 version_0_7_5
General :
 - install Dijon's ChangeLog
Index :
 - files with nested documents (eg mbox) could sometimes not be fully indexed
  and/or and the wrong MIME type was reported
Search :
 - updated Xesam Query Language parser to reflect current spec
 - updated Sherlock plugin for Exalead, removed Ask
Daemon :
 - mbox files are no longer configured separately : those found during
  a crawl are indexed and monitored automatically
 - patterns list can be used as a blacklist (default) or whitelist
UI :
 - fixed a bug where filtering a documents list with an empty query would
  prevent from viewing the list without or without a filter query
 - avoid a crash when viewing the properties of several documents one after
  the other
 - updated Dutch translation by Balaam's Miracle
 - updated Portuguese translation by _PN_boy

2007/06/24 version_0_7_4
General :
 - make uninstall actually uninstalls all files
Index :
 - can detect Hungarian, Romanian and Turkish with libtextcat 2.2 and stem
  with Xapian 1.0
 - adopted Xapian 1.0's new indexing strategy
Search :
 - basic support for the Xesam Query and User Language in pinot-search
Daemon :
 - history database is separate from the UI's
UI :
 - in index list tabs, replaced labels filtering with stored queries filtering,
  so that one can find out which and how many documents in the index being
  shown match a query (the query's maximum number of results is ignored)
 - Hungarian, Romanian and Turkish are valid document languages

2007/05/23 version_0_7_3
Tokenize :
 - mbox filter was broken in previous release
 - TagLib filter returns the file's name as title if no tag is found
SQL :
 - don't force opening and closing the database on every request
Index :
 - overwrite index on upgrades, it's faster than deleting all its documents
 - support for Xapian 1.0 API
Search :
 - stemming wasn't activated when a language is set
 - support for Xapian 1.0 API
Daemon :
 - don't scan for deleted files on every run
UI :
 - fixed crash when unindexing documents
 - updated Dutch translation by JW 
 - updated Italian translation by Marco Bazzani

2007/04/28 version_0_7_2
General :
 - synced with gtk+'s xdgmime
Index :
 - add extra term to documents so that filter "dir:/" can be applied
Search :
 - results lists can be saved as CSV or OpenSearch response XML/RSS
Daemon :
 - indexes created with version < 0.72 will be automatically upgraded
 - fixed problem where too much time spent upgrading would cause the daemon
  to be killed
UI :
 - open documents on HTTP/HTTPS with the default Web browser
 - updated German translation by Christian Dywan

Index/pinot-label.cpp
/*
 *  Copyright 2007 Fabrice Colin
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */
 
#include <getopt.h>
#include <sys/types.h>
#include <pwd.h>
#include <iostream>
#include <string>
#include <set>
#include "config.h"
extern "C"
{
#if DBUS_NUM_VERSION < 1000000
#define DBUS_API_SUBJECT_TO_CHANGE
#endif
#include <dbus/dbus.h>
#include <dbus/dbus-glib.h>
#include <dbus/dbus-glib-lowlevel.h>
}

#include "StringManip.h"
#include "MIMEScanner.h"
#include "DBusXapianIndex.h"
#include "XapianDatabaseFactory.h"

using namespace std;

static struct option g_longOptions[] = {
	{"get", 0, 0, 'g'},
	{"help", 0, 0, 'h'},
	{"list", 0, 0, 'l'},
	{"set", 1, 0, 's'},
	{"version", 0, 0, 'v'},
	{0, 0, 0, 0}
};


static void printLabels(const set<string> &labels, const string &fileName)
{
	if (fileName.empty() == false)
	{
		cout << fileName << endl;
	}
	cout << "Labels: ";

	for (set<string>::const_iterator labelIter = labels.begin();
		labelIter != labels.end(); ++labelIter)
	{
		if (labelIter->substr(0, 2) == "X-")
		{
			continue;
		}
		cout << "[" << Url::escapeUrl(*labelIter) << "]";
	}
	cout << endl;
}

static string getHomeDirectory(void)
{
	struct passwd *pPasswd = getpwuid(geteuid());

	if ((pPasswd != NULL) &&
		(pPasswd->pw_dir != NULL))
	{
		return pPasswd->pw_dir;
	}
	else
	{
		char *homeDir = getenv("HOME");
		if (homeDir != NULL)
		{
			return homeDir;
		}
	}

	return "~";
}

static void printHelp(void)
{
	// Help
	cout << "pinot-label - Label files from the command-line\n\n"
		<< "Usage: pinot-label [OPTIONS] [FILES]\n\n"
		<< "Options:\n"
		<< "  -g, --get                 get the labels list for the given file\n"
		<< "  -h, --help                display this help and exit\n"
		<< "  -l, --list                list known labels\n"
		<< "  -s, --set                 set labels on the given file\n"
		<< "  -v, --version             output version information and exit\n\n";
	cout << "Examples:\n"
		<< "pinot-label --get /home/fabrice/Documents/Bozo.txt\n\n"
		<< "pinot-label --list\n\n"
		<< "pinot-label --set \"[Clowns][Fun][My Hero]\" /home/fabrice/Documents/Bozo.txt\n\n"
		<< "Report bugs to " << PACKAGE_BUGREPORT << endl;
}

int main(int argc, char **argv)
{
	set<string> labels;
	string labelsString;
	int longOptionIndex = 0;
	unsigned int docId = 0;
	int minArgNum = 1;
	bool getLabels = false, getDocumentLabels = false, setDocumentLabels = false, success = false;

	// Look at the options
	int optionChar = getopt_long(argc, argv, "ghls:v", g_longOptions, &longOptionIndex);
	while (optionChar != -1)
	{
		set<string> engines;

		switch (optionChar)
		{
			case 'g':
				getDocumentLabels = true;
				break;
			case 'h':
				printHelp();
				return EXIT_SUCCESS;
			case 'l':
				minArgNum = 0;
				getLabels = true;
				break;
			case 's':
				setDocumentLabels = true;
				if (optarg != NULL)
				{
					labelsString = optarg;
				}
				break;
			case 'v':
				cout << "pinot-label - " << PACKAGE_STRING << "\n\n"
					<< "This is free software.  You may redistribute copies of it under the terms of\n"
					<< "the GNU General Public License <http://www.gnu.org/licenses/old-licenses/gpl-2.0.html>.\n"
					<< "There is NO WARRANTY, to the extent permitted by law." << endl;
				return EXIT_SUCCESS;
			default:
				return EXIT_FAILURE;
		}

		// Next option
		optionChar = getopt_long(argc, argv, "ghls:v", g_longOptions, &longOptionIndex);
	}

	if (argc == 1)
	{
		printHelp();
		return EXIT_SUCCESS;
	}

	if ((argc < 2) ||
		(argc - optind < minArgNum))
	{
		cerr << "Not enough parameters" << endl;
		return EXIT_FAILURE;
	}

	if ((setDocumentLabels == true) &&
		(labelsString.empty() == true))
	{
		cerr << "Incorrect parameters" << endl;
		return EXIT_FAILURE;
	}

	// Initialize GType
	g_type_init();

	MIMEScanner::initialize("", "");

	string indexLocation(getHomeDirectory() + "/.pinot/daemon");
	DBusXapianIndex index(indexLocation);
	if (index.isGood() == false)
	{
		cerr << "Couldn't obtain index for " << indexLocation << endl;

		XapianDatabaseFactory::closeAll();
		MIMEScanner::shutdown();

		return EXIT_FAILURE;
	}

	if (getLabels == true)
	{
		if (index.getLabels(labels, true) == true)
		{
			printLabels(labels, "");

			success = true;
		}
	}

	while (optind < argc)
	{
		string fileParam(argv[optind]);

		if ((getDocumentLabels == true) ||
			(setDocumentLabels == true))
		{
			docId = index.hasDocument(string("file://") + fileParam);
			if (docId == 0)
			{
				cerr << fileParam << " is not indexed" << endl;

				success = false;
				continue;
			}
		}

		if (getDocumentLabels == true)
		{
			labels.clear();

			if (index.getDocumentLabels(docId, labels, true) == true)
			{
				printLabels(labels, fileParam);

				success = true;
			}
		}

		if (setDocumentLabels == true)
		{
			string::size_type endPos = 0;
			string label(StringManip::extractField(labelsString, "[", "]", endPos));

			labels.clear();

			// Parse labels
			while (label.empty() == false)
			{
				labels.insert(Url::unescapeUrl(label));

				if (endPos == string::npos)
				{
					break;
				}
				label = StringManip::extractField(labelsString, "[", "]", endPos);
			}

#ifdef DEBUG
			printLabels(labels, fileParam);
#endif
			success = index.setDocumentLabels(docId, labels);
		}

		// Next
		++optind;
	}

	XapianDatabaseFactory::closeAll();
	MIMEScanner::shutdown();

	// Did whatever operation we carried out succeed ?
	if (success == true)
	{
		return EXIT_SUCCESS;
	}

	return EXIT_FAILURE;
}
